man/rbd-replay.8
.\" Man page generated from reStructuredText.
.
.TH "RBD-REPLAY" "8" "February 04, 2016" "dev" "Ceph"
.SH NAME
rbd-replay \- replay rados block device (RBD) workloads
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.nf
\fBrbd\-replay\fP [ \fIoptions\fP ] \fIreplay_file\fP
.fi
.sp
.SH DESCRIPTION
.sp
\fBrbd\-replay\fP is a utility for replaying rados block device (RBD) workloads.
.SH OPTIONS
.INDENT 0.0
.TP
.B \-c ceph.conf, \-\-conf ceph.conf
Use ceph.conf configuration file instead of the default /etc/ceph/ceph.conf to
determine monitor addresses during startup.
.UNINDENT
.INDENT 0.0
.TP
.B \-p pool, \-\-pool pool
Interact with the given pool.  Defaults to \(aqrbd\(aq.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-latency\-multiplier
Multiplies inter\-request latencies.  Default: 1.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-read\-only
Only replay non\-destructive requests.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-map\-image rule
Add a rule to map image names in the trace to image names in the replay cluster.
A rule of image1@snap1=image2@snap2 would map snap1 of image1 to snap2 of image2.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-dump\-perf\-counters
\fBExperimental\fP
Dump performance counters to standard out before an image is closed.
Performance counters may be dumped multiple times if multiple images are closed,
or if the same image is opened and closed multiple times.
Performance counters and their meaning may change between versions.
.UNINDENT
.SH EXAMPLES
.sp
To replay workload1 as fast as possible:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
rbd\-replay \-\-latency\-multiplier=0 workload1
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
To replay workload1 but use test_image instead of prod_image:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
rbd\-replay \-\-map\-image=prod_image=test_image workload1
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AVAILABILITY
.sp
\fBrbd\-replay\fP is part of Ceph, a massively scalable, open\-source, distributed storage system. Please refer to
the Ceph documentation at \fI\%http://ceph.com/docs\fP for more information.
.SH SEE ALSO
.sp
\fBrbd\-replay\-prep\fP(8),
\fBrbd\fP(8)
.SH COPYRIGHT
2010-2014, Inktank Storage, Inc. and contributors. Licensed under Creative Commons BY-SA
.\" Generated by docutils manpage writer.
.


src/rbd_replay/rbd-replay.cc
// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
// vim: ts=8 sw=2 smarttab
/*
 * Ceph - scalable distributed file system
 *
 * Copyright (C) 2014 Adam Crume <adamcrume@gmail.com>
 *
 * This is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1, as published by the Free Software
 * Foundation.  See file COPYING.
 *
 */

#include <vector>
#include <thread>
#include "common/ceph_argparse.h"
#include "global/global_init.h"
#include "Replayer.hpp"
#include "rbd_replay_debug.hpp"
#include "ImageNameMap.hpp"


using namespace std;
using namespace rbd_replay;


static const char* get_remainder(const char *string, const char *prefix) {
  while (*prefix) {
    if (*prefix++ != *string++) {
      return NULL;
    }
  }
  return string;
}

static void usage(const char* program) {
  cout << "Usage: " << program << " --conf=<config_file> <replay_file>" << std::endl;
  cout << "Options:" << std::endl;
  cout << "  -p, --pool-name <pool>          Name of the pool to use.  Default: rbd" << std::endl;
  cout << "  --latency-multiplier <float>    Multiplies inter-request latencies.  Default: 1" << std::endl;
  cout << "  --read-only                     Only perform non-destructive operations." << std::endl;
  cout << "  --map-image <rule>              Add a rule to map image names in the trace to" << std::endl;
  cout << "                                  image names in the replay cluster." << std::endl;
  cout << "  --dump-perf-counters            *Experimental*" << std::endl;
  cout << "                                  Dump performance counters to standard out before" << std::endl;
  cout << "                                  an image is closed. Performance counters may be dumped" << std::endl;
  cout << "                                  multiple times if multiple images are closed, or if" << std::endl;
  cout << "                                  the same image is opened and closed multiple times." << std::endl;
  cout << "                                  Performance counters and their meaning may change between" << std::endl;
  cout << "                                  versions." << std::endl;
  cout << std::endl;
  cout << "Image mapping rules:" << std::endl;
  cout << "A rule of image1@snap1=image2@snap2 would map snap1 of image1 to snap2 of" << std::endl;
  cout << "image2." << std::endl;
}

int main(int argc, const char **argv) {
  vector<const char*> args;

  argv_to_vec(argc, argv, args);
  env_to_vec(args);
  global_init(NULL, args, CEPH_ENTITY_TYPE_CLIENT, CODE_ENVIRONMENT_UTILITY, 0);

  std::vector<const char*>::iterator i;
  string pool_name = "rbd";
  float latency_multiplier = 1;
  bool readonly = false;
  ImageNameMap image_name_map;
  std::string val;
  std::ostringstream err;
  bool dump_perf_counters = false;
  for (i = args.begin(); i != args.end(); ) {
    if (ceph_argparse_double_dash(args, i)) {
      break;
    } else if (ceph_argparse_witharg(args, i, &val, "-p", "--pool", (char*)NULL)) {
      pool_name = val;
    } else if (ceph_argparse_witharg(args, i, &latency_multiplier, err, "--latency-multiplier",
				     (char*)NULL)) {
      if (!err.str().empty()) {
	cerr << err.str() << std::endl;
	return 1;
      }
    } else if (ceph_argparse_flag(args, i, "--read-only", (char*)NULL)) {
      readonly = true;
    } else if (ceph_argparse_witharg(args, i, &val, "--map-image", (char*)NULL)) {
      ImageNameMap::Mapping mapping;
      if (image_name_map.parse_mapping(val, &mapping)) {
	image_name_map.add_mapping(mapping);
      } else {
	cerr << "Unable to parse mapping string: '" << val << "'" << std::endl;
	return 1;
      }
    } else if (ceph_argparse_flag(args, i, "-h", "--help", (char*)NULL)) {
      usage(argv[0]);
      return 0;
    } else if (ceph_argparse_flag(args, i, "--dump-perf-counters", (char*)NULL)) {
      dump_perf_counters = true;
    } else if (get_remainder(*i, "-")) {
      cerr << "Unrecognized argument: " << *i << std::endl;
      return 1;
    } else {
      ++i;
    }
  }

  common_init_finish(g_ceph_context);

  string replay_file;
  if (!args.empty()) {
    replay_file = args[0];
  }

  if (replay_file.empty()) {
    cerr << "No replay file specified." << std::endl;
    return 1;
  }

  unsigned int nthreads = std::thread::hardware_concurrency();
  Replayer replayer(2 * nthreads + 1);
  replayer.set_latency_multiplier(latency_multiplier);
  replayer.set_pool_name(pool_name);
  replayer.set_readonly(readonly);
  replayer.set_image_name_map(image_name_map);
  replayer.set_dump_perf_counters(dump_perf_counters);
  replayer.run(replay_file);
}
